app/src/studio/uart_rpc_transport.c

```diff
@@ -0,0 +1,167 @@
+
+LOG_MODULE_DECLARE(zmk_studio, CONFIG_ZMK_STUDIO_LOG_LEVEL);
+
+/* change this to any other UART peripheral if desired */
+#define UART_DEVICE_NODE DT_CHOSEN(zmk_studio_rpc_uart)
+
+static const struct device *const uart_dev = DEVICE_DT_GET(UART_DEVICE_NODE);
+
+static void tx_notify(struct ring_buf *tx_ring_buf, size_t written, bool msg_done,
+                      void *user_data) {
+    if (msg_done || (ring_buf_size_get(tx_ring_buf) > (ring_buf_capacity_get(tx_ring_buf) / 2))) {
+#if IS_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN)
+        uart_irq_tx_enable(uart_dev);
+#else
+        struct ring_buf *tx_buf = zmk_rpc_get_tx_buf();
+        uint8_t *buf;
+        uint8_t claim_len;
+        while ((claim_len = ring_buf_get_claim(tx_buf, &buf, tx_buf->size)) > 0) {
+            for (int i = 0; i < claim_len; i++) {
+                uart_poll_out(uart_dev, buf[i]);
+            }
+
+            ring_buf_get_finish(tx_buf, claim_len);
+        }
+#endif
+    }
+}
+
+#if !IS_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN)
+
+static void uart_rx_main(void) {
+    for (;;) {
+        uint8_t *buf;
+        struct ring_buf *ring_buf = zmk_rpc_get_rx_buf();
+        uint8_t claim_len = ring_buf_put_claim(ring_buf, &buf, 1);
+
+        if (claim_len < 1) {
+            LOG_WRN("NO CLAIM ABLE TO BE HAD");
+            k_sleep(K_MSEC(1));
+            continue;
+        }
+
+        if (uart_poll_in(uart_dev, buf) < 0) {
+            ring_buf_put_finish(ring_buf, 0);
+            k_sleep(K_MSEC(1));
+        } else {
+            ring_buf_put_finish(ring_buf, 1);
+            zmk_rpc_rx_notify();
+        }
+    }
+}
+
+K_THREAD_DEFINE(uart_transport_read_thread, CONFIG_ZMK_STUDIO_TRANSPORT_UART_RX_STACK_SIZE,
+                uart_rx_main, NULL, NULL, NULL, K_LOWEST_APPLICATION_THREAD_PRIO, 0, 0);
+
+#endif
+
+static int start_rx() {
+#if IS_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN)
+    uart_irq_rx_enable(uart_dev);
+#else
+    k_thread_resume(uart_transport_read_thread);
+#endif
+    return 0;
+}
+
+static int stop_rx(void) {
+#if IS_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN)
+    uart_irq_rx_disable(uart_dev);
+#else
+    k_thread_suspend(uart_transport_read_thread);
+#endif
+    return 0;
+}
+
+ZMK_RPC_TRANSPORT(uart, ZMK_TRANSPORT_USB, start_rx, stop_rx, NULL, tx_notify);
+
+#if IS_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN)
+
+/*
+ * Read characters from UART until line end is detected. Afterwards push the
+ * data to the message queue.
+ */
+static void serial_cb(const struct device *dev, void *user_data) {
+    if (!uart_irq_update(uart_dev)) {
+        return;
+    }
+
+    if (uart_irq_rx_ready(uart_dev)) {
+        /* read until FIFO empty */
+        uint32_t last_read = 0;
+        struct ring_buf *buf = zmk_rpc_get_rx_buf();
+        do {
+            uint8_t *buffer;
+            uint32_t len = ring_buf_put_claim(buf, &buffer, buf->size);
+            if (len == 0) {
+                zmk_rpc_rx_notify();
+                continue;
+            }
```

Comment:
Does this just busy loop until it gets a claim? If so, I think this would need to be running on a thread with a lower priority than whatever is consuming the data, or it may deadlock by never allowing that thread to run?

Reply:
A good question... The `zmk_rpc_rx_notify` will give on the underlying semaphore, but this is in an ISR context for the UART callback, IIRC, so we can't yield... I'm honestly not sure if we're hitting this (yet) in practice, and if so, how it's functioning. Let me dig a bit.

Reply:
Ugh... Ok, so... I think the answer is: We need to size the RX ring buffer large enough to avoid this happening. From https://docs.zephyrproject.org/3.5.0/doxygen/html/group__uart__interrupt.html#gab10942076ac47ecff29e924098049398:

> That means that once [uart_irq_rx_ready()](https://docs.zephyrproject.org/3.5.0/doxygen/html/group__uart__interrupt.html#gad04073b1b8e3de13b43ae1194561377b) is detected, [uart_fifo_read()](https://docs.zephyrproject.org/3.5.0/doxygen/html/group__uart__interrupt.html#gab10942076ac47ecff29e924098049398) must be called until it reads all available data in the FIFO (i.e. until it returns less data than was requested).

So... we *have* to keep calling `uart_fifo_read` from this callback, even if we don't have anywhere to put it... And we're an an ISR so we can't just yield. I think the best we can actually do in this scenario is skip the bytes `LOG_ERR`
that we're doing so, and bail.

Thoughts?

Reply:
Yup, here's a pretty authoritative Zephyr use of the async API for the shell work. It basically drops the data if there's no room in the ring buffer:

https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/shell/backends/shell_uart.c#L84

(modified to remove some conditional code for clarity)

```
	do {
		len = ring_buf_put_claim(&sh_uart->rx_ringbuf, &data,
					 sh_uart->rx_ringbuf.size);

		if (len > 0) {
			rd_len = uart_fifo_read(dev, data, len);

			/* If there is any new data to be either taken into
			 * ring buffer or consumed by the SMP, signal the
			 * shell_thread.
			 */
			if (rd_len > 0) {
				new_data = true;
			}

			int err = ring_buf_put_finish(&sh_uart->rx_ringbuf, rd_len);
			(void)err;
			__ASSERT_NO_MSG(err == 0);
		} else {
			uint8_t dummy;

			/* No space in the ring buffer - consume byte. */
			LOG_WRN("RX ring buffer full.");

			rd_len = uart_fifo_read(dev, &dummy, 1);
		}
	} while (rd_len && (rd_len == len));
```

Reply:
Updated to match.